File: norton-embroidery/src/processing/stitches/running.ts
/** Running, bean and outline stitch generation. */

import { densifyPath, distance, type Point, type Ring } from '../../domain/geometry';

export interface RunningOptions {
  stitchLength: number;
  /** Drop points closer together than this (0.1 mm units). */
  minStitchLength: number;
}

/** Even-spaced needle penetrations along a polyline. */
export function runningStitch(path: readonly Point[], options: RunningOptions): Point[] {
  if (path.length < 2) return path.length ? [{ ...path[0] }] : [];
  const dense = densifyPath(path, options.stitchLength);
  return dropShortSegments(dense, options.minStitchLength);
}

/** Running stitch around a closed ring, returning to the start point. */
export function outlineStitch(ring: Ring, options: RunningOptions): Point[] {
  if (ring.length < 2) return [];
  return runningStitch([...ring, ring[0]], options);
}

/**
 * Bean stitch: the same path sewn back and forth `repeats` times, giving a
 * heavier line. `repeats` must be odd so the needle finishes at the far end.
 */
export function beanStitch(
  path: readonly Point[],
  repeats: number,
  options: RunningOptions,
): Point[] {
  const base = runningStitch(path, options);
  if (base.length < 2) return base;
  const passes = Math.max(1, repeats % 2 === 0 ? repeats + 1 : repeats);
  const out: Point[] = [...base];
  for (let p = 1; p < passes; p++) {
    const source = p % 2 === 1 ? [...base].reverse() : base;
    // Skip the first point of each pass: the needle is already there.
    for (let i = 1; i < source.length; i++) out.push({ ...source[i] });
  }
  return out;
}

/**
 * Remove intermediate points that sit closer than `minLength` to the previous
 * kept point. Very short stitches cause thread breaks and needle damage.
 * Endpoints are always preserved.
 */
export function dropShortSegments(points: readonly Point[], minLength: number): Point[] {
  if (points.length <= 2 || minLength <= 0) return [...points];
  const out: Point[] = [points[0]];
  for (let i = 1; i < points.length - 1; i++) {
    if (distance(out[out.length - 1], points[i]) >= minLength) out.push(points[i]);
  }
  const last = points[points.length - 1];
  if (distance(out[out.length - 1], last) < minLength && out.length > 1) out.pop();
  out.push(last);
  return out;
}

/** Split any segment longer than `maxLength` into equal sub-stitches. */
export function enforceMaxStitchLength(points: readonly Point[], maxLength: number): Point[] {
  if (points.length < 2 || maxLength <= 0) return [...points];
  return densifyPath(points, maxLength);
}

File: embroidery-genie-ai/backend/app/embroidery/running.py
"""Running stitch, bean stitch and travel-path generation."""

from __future__ import annotations

from dataclasses import dataclass

from .geometry import Point, dedupe, distance, resample_path


@dataclass(slots=True)
class RunParams:
    stitch_length: float = 25.0   # 2.5 mm
    repeats: int = 1              # 3 = bean stitch (there, back, there)
    keep_corners: bool = True


def generate_running(path: list[Point], params: RunParams) -> list[Point]:
    """Even running stitch along a path."""
    pts = resample_path(dedupe(path, 0.3), params.stitch_length, params.keep_corners)
    if len(pts) < 2:
        return []
    if params.repeats <= 1:
        return pts
    out: list[Point] = list(pts)
    for i in range(1, params.repeats):
        leg = list(reversed(pts)) if i % 2 == 1 else list(pts)
        out.extend(leg[1:])
    return out


def generate_bean(path: list[Point], stitch_length: float = 25.0) -> list[Point]:
    """Triple-pass bean stitch: heavier outline without changing colour."""
    return generate_running(path, RunParams(stitch_length=stitch_length, repeats=3))


def close_path(path: list[Point]) -> list[Point]:
    if len(path) > 2 and distance(path[0], path[-1]) > 0.5:
        return list(path) + [path[0]]
    return list(path)


def travel_stitches(start: Point, end: Point, stitch_length: float = 30.0) -> list[Point]:
    """Straight travel run between two points, split into legal stitches."""
    total = distance(start, end)
    if total <= stitch_length:
        return [end]
    return resample_path([start, end], stitch_length, keep_corners=False)[1:]
